src/main/scala/sigmastate/serialization/ValueSerializer.scala

```diff
@@ -100,7 +102,13 @@ object ValueSerializer extends SigmaSerializerCompanion[Value[SType]] {
     AppendSerializer
   ).map(s => (s.opCode, s)).toMap
 
-  def serialize(v: Value[SType]): Array[Byte] = v match {
+  private def serializable(v: Value[SType]): Value[SType] = v match {
+    case upcast: Upcast[SType, _]@unchecked =>
+      upcast.input
```

Comment:
@ergomorphic OTOH, regarding restoring `UpcastSerializer` as a fallback, I don't see a way for Upcast to sneak in through this guard and got serialized.

Reply:
@greenhat  We can think about our IR as instructions of some virtual machine, for which we have an interpreter.
So the question is, whether Upcast is valid instruction or not? 
For every valid instruction we need to have a serializer.
Our nearest goal is to finalize this IR, serialization format and interpreter.
Then we will have to describe all this in specification docs, so alternative implementations would be possible.
We are just giving reference implementation here.

Reply:
@ergomorphic This: "We are just giving reference implementation here." when it all clicked for me. Thanks for the nice write-up!